xds/src/test/java/io/grpc/xds/XdsClientImplTestBase.java

```diff
@@ -3539,7 +3540,9 @@ public void sendingToStoppedServer() throws Exception {
 
       // Send a response and do verifications
       verify(ldsResourceWatcher, never()).onResourceDoesNotExist(LDS_RESOURCE);
+      call.setResponseLatch();
       call.sendResponse(LDS, mf.buildWrappedResource(testListenerVhosts), VERSION_1, "0001");
+      assertThat(call.waitForResponseLatch()).isTrue();
```

Comment:
This doesn't look like it's actually doing anything useful because the CountDownLatch is triggered on this thread within the `call.sendResponse()`, right? I think this is only working because it is slowing down the code.

Looking at the other tests, `call.verifyRequest()` looks to be the proper thing to do, but there's no `timeout()` on the mock verify. So how do they work? It looks like it is direct executors everywhere, so the processing should have completed by the time `call.sendResponse()` is returns. So what is special about this test? The connection drops and has to reconnect; maybe there's something there. It seems you already knew there was a problem, as you had the `Thread.sleep(1)` already.

There's clearly another thread involved, otherwise the `Thread.sleep(1)` wouldn't have had an effect. There does seem to be connect backoff involved as the test takes ~1s to complete, and that backoff is most likely occurring within `resourceDiscoveryCalls.poll()`. What is probably happening is the ScheduledExecutorService thread waits 1s, connects, and starts the RPC. But the RPC callbacks use directExecutor on the ScheduledExecutorService thread. That thread is still running when `resourceDiscoveryCall.poll()` returns. We wrap the directExecutor in a SerializingExecutor, so if more callbacks occur before the previous callback completes then those callbacks will remain on the ScheduledExecutorService thread.

That is troublesome. The easiest solution given the current state of the world is to probably pass our own (real) single-threaded ScheduledExecutorService to InProcessChannelBuilder.scheduledExecutorService(). Then, instead of the sleep() we submit() a noop Runnable and wait for the future to complete. When the future completes it means all items queued at the time we submit()ed completed.

Reply:
Moved the countdownLatch call into requestObserver.onNext so that it really is inline with the result processing.  I suspect that the server restarting creates active threads which introduce contention and therefore the appearance of the timing issue.